proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+
+    public static func package(
+        name: String,
+        path: String,
+        traits: Traits
+    ) -> Package.Dependency
+
+    // MARK: Source repository
+
+    public static func package(
+        url: String,
+        from version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        branch: String,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        revision: String,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        _ range: Range<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        _ range: ClosedRange<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        exact version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    // MARK: Registry 
+
+    public static func package(
+        id: String,
+        from version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        exact version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        _ range: Range<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        _ range: ClosedRange<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+}
+```
+
+### Default traits
+
+Default traits allow package authors to define a set of traits that they think
+cater to the majority use-cases of the package. When choosing the initial
+default traits or adding a new default trait it is important to consider that
+removing a default trait is a SemVer major change since it can potentially
+remove APIs.
+
+### Trait specific command line options for `swift build/run`
+
+When executing one of `swift build/run` options can be passed to control
+which traits for the root package are enabled:
+
+- `--traits` _TRAITS_: Enables the passed traits of the package. Multiple traits can be
+  specified by providing a comma separated list e.g. `--traits Trait1,Trait2`.
+- `--enable-all-traits`: Enables all traits of the package.
+- `--disable-default-traits`: Disables all default traits of the package.
+
+
+### Trait limitations
+
+To prevent abuse, limit the complexity and make sure it integrates with the
+compiler a few limitations are imposed.
+
+#### Number of traits
+
+[Other
+ecosystems](https://blog.rust-lang.org/2023/10/26/broken-badges-and-23k-keywords.html)
+have shown that a large number of traits can have significant impact on
+registries and dependency managers. To avoid such a scenario an initial maximum
+number of traits of 300 per package is imposed. This can be revisited later once
```

Comment:
Is this 300 definitions per package or can a package only enable 300 traits across all of its dependencies?

Reply:
I clarified this slightly:

> To avoid such a scenario an initial maximum
number of 300 defined traits per package is imposed